app/services/work_packages/bulk/copy_service.rb

```diff
@@ -32,11 +32,48 @@ class CopyService < BulkedService
       def initialize(user:, work_packages:)
         super
 
+        @wp_map = {}
+
         self.work_packages = remove_hierarchy_duplicates(work_packages)
       end
 
       private
 
+      def bulk(params)
+        result = ServiceResult.success result: true
+
+        work_packages.each do |work_package|
+          # As updating one work package might have already saved another one,
+          # e.g. by changing the start/due date or the version
+          # we need to reload the work packages to avoid running into stale object errors.
+          work_package.reload
+
+          call_move_hook(work_package, params)
+          wp_copy = alter_work_package(work_package, params)
+
+          result.add_dependent!(wp_copy)
+
+          @wp_map.store(work_package.id, wp_copy.result.id)
+        end
+
+        result.result = false if result.failure?
+
+        copy_relations
```

Comment:
I believe the relations should be copied only when the `result` service call is successful. Using the `on_success` method can be a good approach, similar to what we have in the [`WorkPackages::CopyService`](https://github.com/opf/openproject/blob/c427d1ba3514ffe5d2731d9233f5fe87e797c92c/app/services/work_packages/copy_service.rb#L56-L60).